### PR TITLE
Dev/363 Download CSV file for exam's scores

### PR DESCRIPTION
**Pull request by snowmancc**
resolve: #363 

image

### CODE REVIEW

File: Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/exam/CreateExamTranscriptUseCase.java [hunk 1]
```diff
@@ -72,16 +74,14 @@ public void execute(int examId, Presenter presenter) {
...
-    private GetById<Integer, Student> examinees(Exam exam) {
-        var idToExaminee = toMap(studentServiceDriver.getStudentsByIds(
+    private Map<Integer, Student> examinees(Exam exam) {
+        return toMap(studentServiceDriver.getStudentsByIds(
                 mapToList(exam.getExaminees(), Examinee::getStudentId)), Student::getId, identity());
-        return idToExaminee::get;
     }
 
-    private List<ExamineeRecord> examineeRecords(GetById<Integer, Student> examinees, List<QuestionRecord> records) {
+    private List<ExamineeRecord> examineeRecords(Map<Integer, Student> examinees, List<QuestionRecord> records) {
```
Comment (Wally5077) [inline Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/exam/CreateExamTranscriptUseCase.java:82]: NIT: Why do you change ```GetById``` to ```Map```?

File: Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptCsvFilePresenter.java [hunk 2]
```diff
@@ -27,6 +28,7 @@
 public class ExamTranscriptCsvFilePresenter implements CreateExamTranscriptUseCase.Presenter {
     private List<ProblemView> problems;
     private List<ExamineeRecord> examineeRecords;
+    private List<Student> students;
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptCsvFilePresenter.java:31]:
```Java
private List<Student> examinees
```

File: Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptCsvFilePresenter.java [hunk 3]
```diff
@@ -46,13 +48,31 @@ public void showProblems(List<ProblemView> problems) {
         this.problems = problems;
     }
 
+    @Override
+    public void showExaminees(List<Student> students) {
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptCsvFilePresenter.java:52]: ditto

File: Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptCsvFilePresenter.java [hunk 4]
```diff
@@ -46,13 +48,31 @@ public void showProblems(List<ProblemView> problems) {
...
     @SneakyThrows
     public FileResource present() {
         var out = new ByteArrayOutputStream();
         var csvPrinter = new CSVPrinter(new PrintWriter(out, false, UTF_8),
                 CSVFormat.DEFAULT.withHeader(createExamTranscriptCsvHeader(problems)));
 
-        var csvBody = createExamTranscriptCsvBody(examineeRecords);
+        Map<Integer, ExamineeRecord> map = toMap(examineeRecords,
+                examineeRecord -> examineeRecord.getExaminee().getId(),
+                examineeRecord -> examineeRecord);
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptCsvFilePresenter.java:64]:
```Java
var examineeRecords
```

File: Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptCsvFilePresenter.java [hunk 5]
```diff
@@ -46,13 +48,31 @@ public void showProblems(List<ProblemView> problems) {
...
+
+        List<ExamineeRecord> newExamineeRecords = new ArrayList<>();
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptCsvFilePresenter.java:66]:
```Java
List<ExamineeRecord> newExamineeRecords = new ArrayList<>(students.size());
```

File: Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptCsvFilePresenter.java [hunk 6]
```diff
@@ -46,13 +48,31 @@ public void showProblems(List<ProblemView> problems) {
...
+        for (Student student : students) {
+            if (map.containsKey(student.getId())) {
+                newExamineeRecords.add(map.get(student.getId()));
+            } else {
+                newExamineeRecords.add(new ExamineeRecord(student, new ArrayList<>()));
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptCsvFilePresenter.java:71]:
```Java
newExamineeRecords.add(new ExamineeRecord(student,EMPTY_LIST));
```

File: Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptCsvFilePresenter.java [hunk 7]
```diff
@@ -46,13 +48,31 @@ public void showProblems(List<ProblemView> problems) {
...
+            }
+        }
+
+        var csvBody = createExamTranscriptCsvBody(newExamineeRecords);
         for (List<String> record : csvBody) {
             csvPrinter.printRecord(record);
         }
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptCsvFilePresenter.java:78]:
```Java
csvBody.foreach(csvPrinter:: printRecord);
```

File: Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptPresenter.java [hunk 8]
```diff
@@ -39,6 +40,10 @@ public void showRecords(List<ExamineeRecord> examineeRecords) {
     public void showProblems(List<ProblemView> problems) {
     }
 
+    @Override
+    public void showExaminees(List<Student> students) {
+    }
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptPresenter.java:45]:
```Java
List<Student> examinees
```

**Review by Wally5077 — CHANGES_REQUESTED** [review]
Review done

File: Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/exam/CreateExamTranscriptUseCase.java [hunk 9]
```diff
@@ -99,6 +99,8 @@ public void execute(int examId, Presenter presenter) {
         void showRecords(List<ExamineeRecord> examineeRecords);
 
         void showProblems(List<ProblemView> problems);
+
+        void showExaminees(List<Student> students);
```
Comment (Wally5077) [inline Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/exam/CreateExamTranscriptUseCase.java:103]:
```Java
void showExaminees(List<Student> examinees);
```

**Review by Wally5077 — CHANGES_REQUESTED** [review]

File: Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptCsvFilePresenter.java [hunk 10]
```diff
@@ -46,13 +48,31 @@ public void showProblems(List<ProblemView> problems) {
...
+                examineeRecord -> examineeRecord.getExaminee().getId(),
+                examineeRecord -> examineeRecord);
+
+        List<ExamineeRecord> newExamineeRecords = new ArrayList<>();
+        for (Student student : students) {
+            if (map.containsKey(student.getId())) {
+                newExamineeRecords.add(map.get(student.getId()));
+            } else {
+                newExamineeRecords.add(new ExamineeRecord(student, new ArrayList<>()));
+            }
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptCsvFilePresenter.java:72]:
```java
var examineeRecord = examineeRecords.computeIfAbsent(student.getId(), id -> new ExamineeRecord(student, EMPTY_LIST));
newExamineeRecords.add(examineeRecord);
```

File: Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/exam/CreateExamTranscriptUseCase.java (hunk 1 above)
Comment (Johnny850807) [reply Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/exam/CreateExamTranscriptUseCase.java:82]: @snowmancc WHY?

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
Review done

File: Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/exam/CreateExamTranscriptUseCase.java (hunk 1 above)
Comment (snowmancc) [reply Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/exam/CreateExamTranscriptUseCase.java:82]: I need to set the student list into the presenter, So I need the map to use them achieved goal~

Comment (Johnny850807) [reply Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/exam/CreateExamTranscriptUseCase.java:82]: got it

**Review by Johnny850807 — APPROVED** [review]
LGTM

**Review by Wally5077 — APPROVED** [review]
LGTM

**Pull request merged**